pkg/controller/direct/bigquerydatatransfer/bigquerydatatransferconfig_controller.go

```diff
@@ -0,0 +1,351 @@
+	}
+	// TODO(jingyih): need to figure out how to deal with resourceID. It is service generated.
+
+	location := obj.Spec.Location
+	if location == "" {
+		return nil, fmt.Errorf("cannot resolve location")
+	}
+
+	// Resolve PubSubTopic Ref
+	if obj.Spec.PubSubTopicRef != nil {
+		topic, err := refv1beta1.ResolvePubSubTopic(ctx, reader, obj, obj.Spec.PubSubTopicRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.PubSubTopicRef.External = topic.String()
+	}
+
+	// Resolve BigQueryDataSet Ref
+	if obj.Spec.DatasetRef != nil {
+		dataset, err := refv1beta1.ResolveBigQueryDataset(ctx, reader, obj, obj.Spec.DatasetRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.DatasetRef.External = dataset.String()
+	}
+
+	// Resolve KMSCryptoKey Ref
+	if obj.Spec.EncryptionConfiguration != nil {
+		kmsCryptoKeyRef, err := refv1beta1.ResolveKMSCryptoKeyRef(ctx, reader, obj, obj.Spec.EncryptionConfiguration.KmsKeyRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.EncryptionConfiguration.KmsKeyRef = kmsCryptoKeyRef
+	}
+
+	// Resolve Project Ref
+	projectRef, err := refv1beta1.ResolveProject(ctx, reader, obj, obj.Spec.ProjectRef)
+	if err != nil {
+		return nil, err
+	}
+	projectID := projectRef.ProjectID
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve project")
+	}
+
+	var id *BigQueryDataTransferConfigIdentity
+	externalRef := direct.ValueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id = BuildID(projectID, location, resourceID)
+	} else {
+		id, err = asID(externalRef)
+		if err != nil {
+			return nil, err
+		}
+
+		if id.projectID != projectID {
+			return nil, fmt.Errorf("BigQueryDataTransferConfig %s/%s has spec.projectRef changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.projectID, projectID)
+		}
+		if id.location != location {
+			return nil, fmt.Errorf("BigQueryDataTransferConfig %s/%s has spec.location changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.location, location)
+		}
+		if id.transferConfigID != resourceID {
+			return nil, fmt.Errorf("BigQueryDataTransferConfig %s/%s has metadata.name or spec.resourceID changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.transferConfigID, resourceID)
+		}
+	}
+
+	// Get bigquerydatatransfer GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func (m *model) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id        *BigQueryDataTransferConfigIdentity
+	gcpClient *gcp.Client
+	desired   *krm.BigQueryDataTransferConfig
+	actual    *bigquerydatatransferpb.TransferConfig
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+
+	if a.id == nil {
+		return false, nil
+	}
```

Comment:
Mind explaining what this means?

Reply:
I will remove this block. I think it was from my previous iteration on how to handle the service generated ID. Thanks for spotting!